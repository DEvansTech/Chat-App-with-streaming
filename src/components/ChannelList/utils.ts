import type {
  Channel,
  LiteralStringForUnion,
  StreamChat,
  UnknownType,
} from 'stream-chat';

type MoveParameters<
  AttachmentType extends UnknownType = UnknownType,
  ChannelType extends UnknownType = UnknownType,
  CommandType extends string = LiteralStringForUnion,
  EventType extends UnknownType = UnknownType,
  MessageType extends UnknownType = UnknownType,
  ReactionType extends UnknownType = UnknownType,
  UserType extends UnknownType = UnknownType
> = {
  channels: Channel<
    AttachmentType,
    ChannelType,
    CommandType,
    EventType,
    MessageType,
    ReactionType,
    UserType
  >[];
  cid: string;
};

export const moveChannelUp = <
  AttachmentType extends UnknownType = UnknownType,
  ChannelType extends UnknownType = UnknownType,
  CommandType extends string = LiteralStringForUnion,
  EventType extends UnknownType = UnknownType,
  MessageType extends UnknownType = UnknownType,
  ReactionType extends UnknownType = UnknownType,
  UserType extends UnknownType = UnknownType
>({
  cid,
  channels = [],
}: MoveParameters<
  AttachmentType,
  ChannelType,
  CommandType,
  EventType,
  MessageType,
  ReactionType,
  UserType
>) => {
  // get channel from channels
  const index = channels.findIndex((c) => c.cid === cid);
  if (index <= 0) return channels;
  const channel = channels[index];

  // remove channel from current position and add to start
  channels.splice(index, 1);
  channels.unshift(channel);

  return [...channels];
};

type GetParameters<
  AttachmentType extends UnknownType = UnknownType,
  ChannelType extends UnknownType = UnknownType,
  CommandType extends string = LiteralStringForUnion,
  EventType extends UnknownType = UnknownType,
  MessageType extends UnknownType = UnknownType,
  ReactionType extends UnknownType = UnknownType,
  UserType extends UnknownType = UnknownType
> = {
  client: StreamChat<
    AttachmentType,
    ChannelType,
    CommandType,
    EventType,
    MessageType,
    ReactionType,
    UserType
  >;
  id: string;
  type: string;
};

export const getChannel = async <
  AttachmentType extends UnknownType = UnknownType,
  ChannelType extends UnknownType = UnknownType,
  CommandType extends string = LiteralStringForUnion,
  EventType extends UnknownType = UnknownType,
  MessageType extends UnknownType = UnknownType,
  ReactionType extends UnknownType = UnknownType,
  UserType extends UnknownType = UnknownType
>({
  client,
  id,
  type,
}: GetParameters<
  AttachmentType,
  ChannelType,
  CommandType,
  EventType,
  MessageType,
  ReactionType,
  UserType
>) => {
  const channel = client.channel(type, id);
  await channel.watch();
  return channel;
};

export const DEFAULT_QUERY_CHANNELS_LIMIT = 10;
export const MAX_QUERY_CHANNELS_LIMIT = 30;
