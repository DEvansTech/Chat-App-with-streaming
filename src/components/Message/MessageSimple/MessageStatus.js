import React from 'react';
import PropTypes from 'prop-types';

import styled from '@stream-io/styled-components';

import loadingGif from '../../../images/loading.gif';
import iconDeliveredUnseen from '../../../images/icons/delivered_unseen.png';

import { Avatar } from '../../Avatar';

const Spacer = styled.View`
  height: 10;
`;

const StatusContainer = styled.View`
  width: 20;
  flex-direction: row;
  justify-content: center;
`;

const DeliveredContainer = styled.View`
  display: flex;
  align-items: center;
  height: 20;
  ${({ theme }) => theme.message.status.deliveredContainer.css};
`;

const DeliveredCircle = styled.View`
  width: 16;
  height: 16;
  border-radius: 16;
  background-color: ${({ theme }) => theme.colors.primary};
  align-items: center;
  justify-content: center;
  padding: 6px;
  ${({ theme }) => theme.message.status.deliveredCircle.css};
`;

const CheckMark = styled.Image`
  width: 8;
  height: 6;
  ${({ theme }) => theme.message.status.checkMark.css};
`;

const SendingContainer = styled.View`
  display: flex;
  align-items: center;
  ${({ theme }) => theme.message.status.sendingContainer.css};
`;

const SendingImage = styled.View`
  height: 10;
  width: 10;
  ${({ theme }) => theme.message.status.sendingImage.css};
`;

const ReadByContainer = styled.View`
  display: flex;
  flex-direction: row;
  align-items: center;
  ${({ theme }) => theme.message.status.readByContainer.css};
`;

const MessageStatus = ({
  client,
  readBy,
  message,
  lastReceivedId,
  threadList,
}) => {
  const renderStatus = () => {
    const justReadByMe =
      readBy.length === 1 && readBy[0] && readBy[0].id === client.user.id;

    if (message.status === 'sending') {
      return (
        <SendingContainer>
          <SendingImage source={loadingGif} />
        </SendingContainer>
      );
    } else if (
      readBy.length !== 0 &&
      !threadList &&
      message.id === lastReceivedId &&
      !justReadByMe
    ) {
      const lastReadUser = readBy.filter(
        (item) => item.id !== client.user.id,
      )[0];
      return (
        <ReadByContainer>
          <Avatar
            name={lastReadUser.name || lastReadUser.id}
            image={lastReadUser.image}
            size={16}
          />
        </ReadByContainer>
      );
    } else if (
      message.status === 'received' &&
      message.type !== 'ephemeral' &&
      message.id === lastReceivedId &&
      !threadList
    ) {
      return (
        <DeliveredContainer>
          <DeliveredCircle>
            <CheckMark source={iconDeliveredUnseen} />
          </DeliveredCircle>
        </DeliveredContainer>
      );
    } else {
      return <Spacer />;
    }
  };

  return <StatusContainer>{renderStatus()}</StatusContainer>;
};

MessageStatus.propTypes = {
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react-native/#channelcontext) */
  client: PropTypes.object,
  /** A list of users who have read the message */
  readBy: PropTypes.array,
  /** Current [message object](https://getstream.io/chat/docs/#message_format) */
  message: PropTypes.object,
  /** Latest message id on current channel */
  lastReceivedId: PropTypes.string,
  /** Boolean if current message is part of thread */
  isThreadList: PropTypes.bool,
};

export default MessageStatus;
