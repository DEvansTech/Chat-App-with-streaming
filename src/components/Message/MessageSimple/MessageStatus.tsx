import React from 'react';

import { Avatar } from '../../Avatar/Avatar';

import { useChatContext } from '../../../contexts/chatContext/ChatContext';
import { styled } from '../../../styles/styledComponents';

import type { ForwardedMessageProps } from './MessageContent';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

const iconDeliveredUnseen = require('../../../images/icons/delivered_unseen.png');
const loadingGif = require('../../../images/loading.gif');

const CheckMark = styled.Image`
  height: 6px;
  width: 8px;
  ${({ theme }) => theme.message.status.checkMark.css};
`;

const DeliveredCircle = styled.View`
  align-items: center;
  background-color: ${({ theme }) => theme.colors.primary};
  border-radius: 16px;
  height: 16px;
  justify-content: center;
  padding: 6px;
  width: 16px;
  ${({ theme }) => theme.message.status.deliveredCircle.css};
`;

const DeliveredContainer = styled.View`
  align-items: center;
  height: 20px;
  ${({ theme }) => theme.message.status.deliveredContainer.css};
`;

const ReadByContainer = styled.View`
  align-items: center;
  flex-direction: row;
  ${({ theme }) => theme.message.status.readByContainer.css};
`;

const SendingContainer = styled.View`
  align-items: center;
  ${({ theme }) => theme.message.status.sendingContainer.css};
`;

const SendingImage = styled.Image`
  height: 10px;
  width: 10px;
  ${({ theme }) => theme.message.status.sendingImage.css};
`;

const Spacer = styled.View`
  height: 10px;
`;

const StatusContainer = styled.View`
  flex-direction: row;
  justify-content: center;
  width: 20px;
`;

export const MessageStatus = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType
>({
  lastReceivedId,
  message,
  readBy = [],
  threadList,
}: ForwardedMessageProps<At, Ch, Co, Ev, Me, Re, Us>) => {
  const { client } = useChatContext<At, Ch, Co, Ev, Me, Re, Us>();
  const justReadByMe = readBy.length === 1 && readBy[0].id === client.user?.id;

  const Status = () => {
    if (message.status === 'sending') {
      return (
        <SendingContainer testID='sending-container'>
          <SendingImage source={loadingGif} />
        </SendingContainer>
      );
    }

    if (
      readBy.length !== 0 &&
      !threadList &&
      message.id === lastReceivedId &&
      !justReadByMe
    ) {
      const lastReadUser = readBy.filter(
        (item) => item.id !== client.user?.id,
      )[0];
      return (
        <ReadByContainer testID='read-by-container'>
          <Avatar
            image={lastReadUser.image}
            name={lastReadUser.name || lastReadUser.id}
            size={16}
          />
        </ReadByContainer>
      );
    }

    if (
      message.status === 'received' &&
      message.type !== 'ephemeral' &&
      message.id === lastReceivedId &&
      !threadList
    ) {
      return (
        <DeliveredContainer testID='delivered-container'>
          <DeliveredCircle>
            <CheckMark source={iconDeliveredUnseen} />
          </DeliveredCircle>
        </DeliveredContainer>
      );
    }

    return <Spacer testID='spacer' />;
  };

  return (
    <StatusContainer>
      <Status />
    </StatusContainer>
  );
};
