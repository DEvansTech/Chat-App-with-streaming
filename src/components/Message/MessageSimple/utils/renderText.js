import React from 'react';
import Markdown from '@stream-io/react-native-simple-markdown';
import anchorme from 'anchorme';
import truncate from 'lodash/truncate';

export const renderText = ({ markdownRules, markdownStyles, message }) => {
  // take the @ mentions and turn them into markdown?
  // translate links
  const { mentioned_users = [], text } = message;

  if (!text) {
    return null;
  }

  let newText = text.trim();
  const urls = anchorme(newText, {
    list: true,
  });

  for (const urlInfo of urls) {
    const displayLink = truncate(urlInfo.encoded.replace(/^(www\.)/, ''), {
      length: 20,
      omission: '...',
    });
    const mkdown = `[${displayLink}](${urlInfo.protocol}${urlInfo.encoded})`;
    newText = newText.replace(urlInfo.raw, mkdown);
  }

  if (mentioned_users.length) {
    for (let i = 0; i < mentioned_users.length; i++) {
      const username = mentioned_users[i].name || mentioned_users[i].id;
      const mkdown = `**@${username}**`;
      const re = new RegExp(`@${username}`, 'g');
      newText = newText.replace(re, mkdown);
    }
  }

  newText = newText.replace(/[<&"'>]/g, '\\$&');
  const styles = {
    ...defaultMarkdownStyles,
    ...markdownStyles,
  };

  return (
    <Markdown rules={markdownRules} styles={styles}>
      {newText}
    </Markdown>
  );
};

const defaultMarkdownStyles = {
  inlineCode: {
    backgroundColor: '#F3F3F3',
    borderColor: '#dddddd',
    color: 'red',
    fontSize: 13,
    padding: 3,
    paddingHorizontal: 5,
  },
  link: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
  url: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
};
