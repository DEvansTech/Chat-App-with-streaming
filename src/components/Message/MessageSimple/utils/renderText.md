[src/components/Message/MessageSimple/utils/renderText.ts](https://github.com/GetStream/stream-chat-react-native/blob/master/src/components/Message/MessageSimple/utils/renderText.ts)

Function that handles markdown and parsing for the text of message.

Internally it uses [react-native-simple-markdown](https://github.com/GetStream/react-native-simple-markdown#styles-1) to handle markdown.

It accepts following params:

- message { object } Message object
- style { object } - <https://github.com/GetStream/react-native-simple-markdown#styles-1>
- markdownRules { object } - <https://github.com/GetStream/react-native-simple-markdown#rules>
