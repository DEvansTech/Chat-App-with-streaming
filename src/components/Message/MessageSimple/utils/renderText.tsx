import React from 'react';
// @ts-expect-error
import Markdown from '@stream-io/react-native-simple-markdown';
import anchorme from 'anchorme';
import truncate from 'lodash/truncate';

import type { MarkdownStyle } from '../../../../styles/themeConstants';
import type { Message } from '../../../MessageList/utils/insertDates';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
} from '../../../../types/types';

type Parameters<
  At extends Record<string, unknown> = DefaultAttachmentType,
  Ch extends Record<string, unknown> = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends Record<string, unknown> = DefaultEventType,
  Me extends Record<string, unknown> = DefaultMessageType,
  Re extends Record<string, unknown> = DefaultReactionType,
  Us extends Record<string, unknown> = DefaultUserType
> = {
  markdownRules: Record<string, unknown>;
  markdownStyles: MarkdownStyle;
  message: Message<At, Ch, Co, Ev, Me, Re, Us>;
};

export const renderText = <
  At extends Record<string, unknown> = DefaultAttachmentType,
  Ch extends Record<string, unknown> = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends Record<string, unknown> = DefaultEventType,
  Me extends Record<string, unknown> = DefaultMessageType,
  Re extends Record<string, unknown> = DefaultReactionType,
  Us extends Record<string, unknown> = DefaultUserType
>({
  markdownRules,
  markdownStyles,
  message,
}: Parameters<At, Ch, Co, Ev, Me, Re, Us>) => {
  // take the @ mentions and turn them into markdown?
  // translate links
  const { mentioned_users = [], text } = message;

  if (!text) {
    return null;
  }

  let newText = text.trim();
  const urls = anchorme(newText, {
    list: true,
  });

  for (const urlInfo of urls) {
    const displayLink = truncate(urlInfo.encoded.replace(/^(www\.)/, ''), {
      length: 20,
      omission: '...',
    });
    const mkdown = `[${displayLink}](${urlInfo.protocol}${urlInfo.encoded})`;
    newText = newText.replace(urlInfo.raw, mkdown);
  }

  if (mentioned_users.length) {
    for (let i = 0; i < mentioned_users.length; i++) {
      const username = mentioned_users[i].name || mentioned_users[i].id;
      const mkdown = `**@${username}**`;
      const re = new RegExp(`@${username}`, 'g');
      newText = newText.replace(re, mkdown);
    }
  }

  newText = newText.replace(/[<&"'>]/g, '\\$&');
  const styles = {
    ...defaultMarkdownStyles,
    ...markdownStyles,
  };

  return (
    <Markdown rules={markdownRules} styles={styles}>
      {newText}
    </Markdown>
  );
};

const defaultMarkdownStyles = {
  inlineCode: {
    backgroundColor: '#F3F3F3',
    borderColor: '#dddddd',
    color: 'red',
    fontSize: 13,
    padding: 3,
    paddingHorizontal: 5,
  },
  link: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
  url: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
};
