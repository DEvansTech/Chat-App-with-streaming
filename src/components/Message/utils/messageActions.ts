import type { MessageType } from '../../MessageList/hooks/useMessageList';

import type { MessageAction } from '../../../contexts/messageOverlayContext/MessageOverlayContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export const messageActions = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>({
  blockUser,
  canModifyMessage,
  copyMessage,
  deleteMessage,
  editMessage,
  error,
  flagMessage,
  isMyMessage,
  isThreadMessage,
  message,
  messageReactions,
  muteUser,
  quotedRepliesEnabled,
  reply,
  retry,
  threadRepliesEnabled,
  threadReply,
}: {
  blockUser: MessageAction | null;
  canModifyMessage: boolean;
  copyMessage: MessageAction | null;
  deleteMessage: MessageAction | null;
  editMessage: MessageAction | null;
  error: boolean;
  flagMessage: MessageAction | null;
  isMyMessage: boolean;
  isThreadMessage: boolean;
  message: MessageType<At, Ch, Co, Ev, Me, Re, Us>;
  messageReactions: boolean;
  muteUser: MessageAction | null;
  reply: MessageAction | null;
  retry: MessageAction | null;
  threadReply: MessageAction | null;
  quotedRepliesEnabled?: boolean;
  threadRepliesEnabled?: boolean;
}): Array<MessageAction | null> | undefined => {
  if (messageReactions) {
    return undefined;
  }

  const actions: Array<MessageAction | null> = [];

  if (error && isMyMessage) {
    actions.push(retry);
  }

  if (quotedRepliesEnabled && !isThreadMessage && !error) {
    actions.push(reply);
  }

  if (threadRepliesEnabled && !isThreadMessage && !error) {
    actions.push(threadReply);
  }

  if (canModifyMessage) {
    actions.push(editMessage);
  }

  if (message.text && !error) {
    actions.push(copyMessage);
  }

  if (!isMyMessage) {
    actions.push(muteUser);
  }

  if (!isMyMessage) {
    actions.push(flagMessage);
  }

  if (!isMyMessage && canModifyMessage) {
    actions.push(blockUser);
  }

  if (canModifyMessage) {
    actions.push(deleteMessage);
  }

  return actions;
};
