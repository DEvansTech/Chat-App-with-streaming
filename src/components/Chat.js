import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { ChatContext, TranslationContext } from '../context';
import { NetInfo } from '../native';

import { themed } from '../styles/theme';
import { Streami18n } from '../utils/Streami18n';
/**
 * Chat - Wrapper component for Chat. The needs to be placed around any other chat components.
 * This Chat component provides the ChatContext to all other components.
 *
 * The ChatContext provides the following props:
 *
 * - client (the client connection)
 * - channels (the list of channels)
 * - setActiveChannel (a function to set the currently active channel)
 * - channel (the currently active channel)
 *
 * It also exposes the withChatContext HOC which you can use to consume the ChatContext
 *
 * @example ./docs/Chat.md
 * @extends PureComponent
 */

export const Chat = themed(
  class Chat extends PureComponent {
    static themePath = '';
    static propTypes = {
      /** The StreamChat client object */
      client: PropTypes.object.isRequired,
      /**
       * Theme object
       *
       * @ref https://getstream.io/chat/react-native-chat/tutorial/#custom-styles
       * */
      style: PropTypes.object,
      logger: PropTypes.func,
      /**
       * Instance of Streami18n class should be provided to Chat component to enable internationalization.
       *
       * Stream provides following list of in-built translations:
       * 1. English (en)
       * 2. Dutch (nl)
       * 3. ...
       * 4. ...
       *
       * Simplest way to start using chat components in one of the in-built languages would be following:
       *
       * ```
       * const i18n = new Streami18n('nl);
       * <Chat client={chatClient} i18nInstance={i18n}>
       *  ...
       * </Chat>
       * ```
       *
       * If you would like to override certain keys in in-built translation.
       * UI will be automatically updated in this case.
       *
       * ```
       * const i18n = new Streami18n('nl);
       *
       * i18n.registerTranslation('nl', {
       *  'Nothing yet...': 'Nog Niet ...',
       *  '{{ firstUser }} and {{ secondUser }} are typing...': '{{ firstUser }} en {{ secondUser }} zijn aan het typen...',
       * });
       *
       * <Chat client={chatClient} i18nInstance={i18n}>
       *  ...
       * </Chat>
       * ```
       *
       * You can use the same function to add whole new language.
       *
       * ```
       * const i18n = new Streami18n('it');
       *
       * i18n.registerTranslation('it', {
       *  'Nothing yet...': 'Non ancora ...',
       *  '{{ firstUser }} and {{ secondUser }} are typing...': '{{ firstUser }} a {{ secondUser }} stanno scrivendo...',
       * });
       *
       * // Make sure to call setLanguage to reflect new language in UI.
       * i18n.setLanguage('it');
       * <Chat client={chatClient} i18nInstance={i18n}>
       *  ...
       * </Chat>
       * ```
       */
      i18nInstance: PropTypes.instanceOf(Streami18n),
    };

    static defaultProps = {
      logger: () => {},
    };

    constructor(props) {
      super(props);

      this.state = {
        // currently active channel
        channel: {},
        isOnline: true,
        connectionRecovering: false,
        t: null,
      };

      this.unsubscribeNetInfo = null;
      this.setConnectionListener();

      this.props.client.on('connection.changed', (event) => {
        if (this._unmounted) return;
        this.setState({
          isOnline: event.online,
          connectionRecovering: !event.online,
        });
      });

      this.props.client.on('connection.recovered', () => {
        if (this._unmounted) return;
        this.setState({ connectionRecovering: false });
      });

      this._unmounted = false;
    }

    async componentDidMount() {
      this.props.logger('Chat component', 'componentDidMount', {
        tags: ['lifecycle', 'chat'],
        props: this.props,
        state: this.state,
      });

      const { i18nInstance } = this.props;

      let streami18n;
      if (i18nInstance && i18nInstance instanceof Streami18n) {
        streami18n = i18nInstance;
      } else {
        streami18n = new Streami18n({ language: 'en' });
      }

      streami18n.registerSetLanguageCallback((t) => {
        this.setState({ t });
      });

      const { t, tDateTimeParser } = await streami18n.getTranslators();
      this.setState({ t, tDateTimeParser });
    }

    componentDidUpdate() {
      this.props.logger('Chat component', 'componentDidUpdate', {
        tags: ['lifecycle', 'chat'],
        props: this.props,
        state: this.state,
      });
    }

    componentWillUnmount() {
      this.props.logger('Chat component', 'componentWillUnmount', {
        tags: ['lifecycle', 'chat'],
        props: this.props,
        state: this.state,
      });

      this._unmounted = true;
      this.props.client.off('connection.recovered');
      this.props.client.off('connection.changed');
      this.props.client.off(this.handleEvent);
      this.unsubscribeNetInfo && this.unsubscribeNetInfo();
    }

    notifyChatClient = (isConnected) => {
      if (this.props.client != null && this.props.client.wsConnection != null) {
        if (isConnected) {
          this.props.client.wsConnection.onlineStatusChanged({
            type: 'online',
          });
        } else {
          this.props.client.wsConnection.onlineStatusChanged({
            type: 'offline',
          });
        }
      }
    };

    setConnectionListener = () => {
      NetInfo.fetch().then((isConnected) => {
        this.notifyChatClient(isConnected);
      });
      this.unsubscribeNetInfo = NetInfo.addEventListener((isConnected) => {
        this.notifyChatClient(isConnected);
      });
    };

    setActiveChannel = (channel) => {
      this.props.logger('Chat component', 'setActiveChannel', {
        tags: ['chat'],
        props: this.props,
        state: this.state,
      });

      if (this._unmounted) return;
      this.setState(() => ({
        channel,
      }));
    };

    getContext = () => ({
      client: this.props.client,
      channel: this.state.channel,
      setActiveChannel: this.setActiveChannel,
      isOnline: this.state.isOnline,
      connectionRecovering: this.state.connectionRecovering,
      logger: this.props.logger,
    });

    render() {
      this.props.logger('Chat component', 'Rerendering', {
        props: this.props,
        state: this.state,
      });

      if (!this.state.t) return null;

      return (
        <ChatContext.Provider value={this.getContext()}>
          <TranslationContext.Provider
            value={{
              t: this.state.t,
              tDateTimeParser: this.state.tDateTimeParser,
            }}
          >
            {this.props.children}
          </TranslationContext.Provider>
        </ChatContext.Provider>
      );
    }
  },
);
