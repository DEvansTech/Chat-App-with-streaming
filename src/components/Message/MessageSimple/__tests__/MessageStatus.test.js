import React from 'react';
import { cleanup, render, waitFor } from '@testing-library/react-native';
import {
  generateMessage,
  generateStaticUser,
  generateUser,
  getTestClientWithUser,
} from 'mock-builders';

import { Chat } from '../../../Chat';
import MessageStatus from '../MessageStatus';
import { Streami18n } from '../../../../utils';

afterEach(cleanup);

describe('MessageStatus', () => {
  it('should render message status with spacer', async () => {
    const id = 'testID';
    const chatClient = await getTestClientWithUser({ id });
    const i18nInstance = new Streami18n();
    const user = generateUser();
    const message = generateMessage({ user });
    const { getByTestId } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus
          client={chatClient}
          lastReceivedId={message.id}
          message={message}
          readBy={[{ id }]}
        />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('spacer')).toBeTruthy();
    });
  });

  it('should render message status with delivered container', async () => {
    const id = 'testID';
    const chatClient = await getTestClientWithUser({ id });
    const i18nInstance = new Streami18n();
    const user = generateUser();
    const message = generateMessage({ user });
    const { getByTestId } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus
          client={chatClient}
          lastReceivedId={message.id}
          message={{ ...message, status: 'received' }}
          readBy={[{ id }]}
        />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('delivered-container')).toBeTruthy();
    });
  });

  it('should render message status with read by container', async () => {
    const id = 'testID';
    const chatClient = await getTestClientWithUser({ id });
    const i18nInstance = new Streami18n();
    const user = generateUser();
    const message = generateMessage({ user });
    const { getByTestId, getByText, rerender, toJSON } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus
          client={chatClient}
          lastReceivedId={message.id}
          message={message}
          readBy={[{ id }, { id: user.id, name: user.name }]}
        />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('avatar-text')).toBeTruthy();
      expect(getByTestId('read-by-container')).toBeTruthy();
      expect(getByText(user.name.charAt(0))).toBeTruthy();
    });

    const staticUser = generateStaticUser(0);
    const staticMessage = generateMessage({ staticUser });

    rerender(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus
          client={chatClient}
          lastReceivedId={staticMessage.id}
          message={staticMessage}
          readBy={[
            { id },
            {
              id: staticUser.id,
              image: staticUser.image,
              name: staticUser.name,
            },
          ]}
        />
      </Chat>,
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
      expect(getByTestId('avatar-image')).toBeTruthy();
      expect(getByTestId('read-by-container')).toBeTruthy();
    });
  });

  it('should render message status with sending container', async () => {
    const i18nInstance = new Streami18n();
    const user = generateUser();
    const chatClient = await getTestClientWithUser({ id: user.id });
    const message = generateMessage({ user });
    const { getByTestId } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus
          client={chatClient}
          message={{ ...message, status: 'sending' }}
          readBy={[]}
        />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('sending-container')).toBeTruthy();
    });
  });
});
