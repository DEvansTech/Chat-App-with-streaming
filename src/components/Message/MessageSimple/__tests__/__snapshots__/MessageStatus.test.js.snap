// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageStatus should render message status with read by container 1`] = `
<View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "width": 20,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          Object {},
        ]
      }
      testID="read-by-container"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
            },
            Object {},
          ]
        }
      >
        <Image
          accessibilityLabel="initials"
          onError={[Function]}
          resizeMethod="resize"
          source={
            Object {
              "uri": "https://images-na.ssl-images-amazon.com/images/I/51t29lLkg8L._AC_SL1000_.jpg",
            }
          }
          style={
            Array [
              Object {
                "borderRadius": 8,
                "height": 16,
                "width": 16,
              },
              Object {},
            ]
          }
          testID="avatar-image"
        />
      </View>
    </View>
  </View>
</View>
`;
